context/src/repo/mod.rs

```diff
@@ -0,0 +1,198 @@
+                bundle_repo_options.repo_head_branch = bundle_repo_options
+                    .repo_head_branch
+                    .or_else(|| git_head.referent_name().map(|s| s.as_bstr().to_string()))
+                    .or_else(|| {
+                        Self::git_head_branch_from_remote_branches(&git_repo)
+                            .ok()
+                            .flatten()
+                    });
+
+                bundle_repo_options.repo_head_sha = bundle_repo_options
+                    .repo_head_sha
+                    .or_else(|| git_head.id().map(|id| id.to_string()));
+
+                if let Ok(commit) = git_head.peel_to_commit_in_place() {
+                    bundle_repo_options.repo_head_commit_epoch = bundle_repo_options
+                        .repo_head_commit_epoch
+                        .or_else(|| commit.time().ok().map(|time| time.seconds));
+                    head_commit_message = commit.message().map(|msg| msg.title.to_string()).ok();
+                    head_commit_author = commit.author().ok().map(|signature| signature.to_owned());
+                }
+            }
+        }
+
+        // Require URL which should be known at this point.
+        let repo_url = bundle_repo_options
+            .repo_url
+            .context("failed to get repo URL")?;
+        let repo_url_parts =
+            RepoUrlParts::from_url(&repo_url).context("failed to parse repo URL")?;
+        let (repo_head_author_name, repo_head_author_email) = head_commit_author
+            .as_ref()
+            .map(|a| (a.name.to_string(), a.email.to_string()))
+            .unwrap_or_default();
+        Ok(BundleRepo {
+            repo: repo_url_parts,
+            repo_root: bundle_repo_options
+                .repo_root
+                .and_then(|p| p.to_str().map(String::from))
+                .unwrap_or_default(),
+            repo_url,
+            repo_head_branch: bundle_repo_options.repo_head_branch.unwrap_or_default(),
+            repo_head_sha: bundle_repo_options.repo_head_sha.unwrap_or_default(),
+            repo_head_commit_epoch: bundle_repo_options
+                .repo_head_commit_epoch
+                .unwrap_or_default(),
+            repo_head_commit_message: head_commit_message.unwrap_or_default(),
+            repo_head_author_name,
+            repo_head_author_email,
+        })
+    }
+
+    fn git_head_branch_from_remote_branches(
+        git_repo: &Repository,
+    ) -> anyhow::Result<Option<String>> {
+        for remote_branch in git_repo
+            .references()?
+            .remote_branches()?
+            .filter_map(Result::ok)
+        {
+            if let Some(target_id) = remote_branch.target().try_id() {
+                if target_id.as_bytes() == remote_branch.id().as_bytes() {
+                    return Ok(remote_branch.name().to_path().to_str().map(String::from));
+                }
+            }
+        }
+        Ok(None)
+    }
+}
```

Comment:
I re-wrote this to be clearer, have less variables getting defined and mutated, and easier to make bindings to for other languages